flang/lib/Lower/CallInterface.cpp

```diff
@@ -486,6 +494,18 @@ class Fortran::lower::CallInterfaceImpl {
     return false;
   }
 
+  mlir::Type
+  translateDynamicType(const Fortran::evaluate::DynamicType &dynamicType) {
+    auto cat = dynamicType.category();
+    if (cat == Fortran::common::TypeCategory::Derived)
+      return getConverter().genType(dynamicType.GetDerivedTypeSpec());
+    llvm::SmallVector<std::int64_t, 2> lenParams;
+    if (cat == Fortran::common::TypeCategory::Character)
+      if (auto constantLen = toInt64(dynamicType.GetCharLength()))
+        lenParams.push_back(*constantLen);
+    return getConverter().genType(cat, dynamicType.kind(), lenParams);
```

Comment:
`lenParams` is only apropos when this is a CHARACTER, right?

Reply:
So far yes, the RecordType does not represent compile time constant length parameters, contrary to fir.char. So there is no need to thread the derived type length parameter values, but only there name and types, which is information accessible through the DerivedTypeSpec.
We may want to make genType lenParams argument a simple optional mlir::Value instead of an ValureRange or ArrayRef. I am waiting that we are deeper into derived types with length parameters to decide how to refactor this.

Reply:
Your comment might also be that it is weird to pass something called `lenParams` for INTEGER, REAL, LOGICAL and COMPLEX. That's OK because it is empty in this case, but I rearranged the code so that it is cleaner and we do not pass a an argument called lenParams for non character.